@types/bun: deprecate some unused types especially on `Bun` ns

Elusive7733 opened the pull request:
### What does this PR do?

<!-- **Please explain what your changes do**, example: -->
Mark unused types as deprecated to improve type clarity and documentation:
- Add deprecation tags to event listener types (BeforeExitListener, etc.)
- Deprecate FileBlob with direction to use BunFile instead
- Replace XMLHttpRequestBodyInit with expanded type definition in fetch.d.ts

<!--

This adds a new flag --bail to bun test. When set, it will stop running tests after the first failure. This is useful for CI environments where you want to fail fast.

-->

- [x] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [ ] Code changes

### How did you verify your code works?

<!-- **For code changes, please include automated tests**. Feel free to uncomment the line below -->

<!-- I wrote automated tests -->

<!-- If JavaScript/TypeScript modules or builtins changed:

- [ ] I included a test for the new code, or existing tests cover it
- [ ] I ran my tests locally and they pass (`bun-debug test test-file-name.test`)

-->

<!-- If Zig files changed:

- [ ] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [ ] I included a test for the new code, or an existing test covers it
- [ ] JSValue used outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [ ] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)
-->

<!-- If new methods, getters, or setters were added to a publicly exposed class:

- [ ] I added TypeScript types for the new methods, getters, or setters
-->

<!-- If dependencies in tests changed:

- [ ] I made sure that specific versions of dependencies are used instead of ranged or tagged versions
-->

<!-- If a new builtin ESM/CJS module was added:

- [ ] I updated Aliases in `module_loader.zig` to include the new module
- [ ] I added a test that imports the module
- [ ] I added a test that require() the module
-->

alii commented on packages/bun-types/fetch.d.ts:
```diff
@@ -12,7 +12,7 @@ declare module "bun" {
   type HeadersInit = string[][] | Record<string, string | ReadonlyArray<string>> | Headers;
   type BodyInit =
     | ReadableStream
-    | Bun.XMLHttpRequestBodyInit
```
Please run prettier on this file

alii reviewed (changes requested):
There are a few more types that should be removed, I think the real fix for this is figuring out a more reliable way to detect these unused types.

Additionally could you please remove the symbols from `bun.d.ts` that you added to `deprecated.d.ts`, otherwise they are declared twice.

Elusive7733 commented:
making changes.

Also we can create a script or that can detect all unused types to deal with this in the future.

alii commented:
Yeah, exactly. A script to detect these would be perfect. I'm not immediately sure how it would behave. Would need some investigation.

Elusive7733 commented:
Hi @alii I have addressed the comments and I am also am working on a script that auto detecs unused types.

I think the PR for the script should be a different one, also there are a few edgcases with the script that needs to be considered.

For now / going forward we should probably add an eslint rule for unused types.

alii commented:
Cool, thanks! I agree, a lint rule would be nice.

Currently busy with some work on the Bun website right now, but I will pull this branch and make sure everything works as expected later today.

Elusive7733 commented:
@alii Are you planning to merge this ?

Also can you point out more good first issues that I can pick up.
Looking forward to contribute to bun.

alii commented on packages/bun-types/deprecated.d.ts:
```diff
@@ -42,6 +45,106 @@ declare module "bun" {
+    | ReadableStreamDefaultReadDoneResult;
+
+  /** @deprecated This type is unused in Bun's types and might be removed in the near future */
+  type MessageEventSource = undefined;
```
It was only used internally in `Bun.MessageEventInit` etc
